File: src/main/java/fr/lewon/bot/runner/bot/operation/def/UpdateBotPropertiesOperation.java
package fr.lewon.bot.runner.bot.operation.def;

import fr.lewon.bot.runner.bot.Bot;
import fr.lewon.bot.runner.bot.operation.BotOperation;
import fr.lewon.bot.runner.bot.operation.OperationResult;
import fr.lewon.bot.runner.bot.props.BotPropertyDescriptor;
import fr.lewon.bot.runner.bot.props.BotPropertyStore;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class UpdateBotPropertiesOperation extends BotOperation {

    public UpdateBotPropertiesOperation() {
        super("Update bot properties");
    }

    @Override
    public List<BotPropertyDescriptor> getNeededProperties(Bot bot) {
        return bot.getBotPropertyStore().keySet().stream()
                .map(bpd -> new BotPropertyDescriptor(
                        bpd.getKey(),
                        bpd.getType(),
                        null,
                        bpd.getDescription(),
                        false,
                        bpd.isNullable(),
                        bpd.getAcceptedValues()))
                .collect(Collectors.toList());
    }

    @Override
    public Object run(Bot bot, BotPropertyStore paramsPropertyStore) throws Exception {

        for (Map.Entry<BotPropertyDescriptor, Object> e : paramsPropertyStore.entrySet()) {
            if (bot.getBotPropertyStore().containsKey(e.getKey())) {
                bot.getBotPropertyStore().put(e.getKey(), e.getValue());
            }
        }
        return new OperationResult(true, "Bot properties updated", bot.getBotPropertyStore());
    }
}

File: src/main/kotlin/fr/lewon/bot/runner/bot/operation/def/UpdateBotPropertiesOperation.kt
package fr.lewon.bot.runner.bot.operation.def

import fr.lewon.bot.runner.Bot
import fr.lewon.bot.runner.bot.operation.BotOperation
import fr.lewon.bot.runner.bot.operation.OperationResult
import fr.lewon.bot.runner.bot.props.BotPropertyDescriptor
import fr.lewon.bot.runner.bot.props.BotPropertyStore
import kotlin.streams.toList

class UpdateBotPropertiesOperation : BotOperation("Update bot properties") {

    override fun getNeededProperties(bot: Bot): List<BotPropertyDescriptor> {
        return bot.botPropertyStore.entries.stream()
            .map {
                BotPropertyDescriptor(
                    it.key.key,
                    it.key.type,
                    it.value,
                    it.key.description,
                    false,
                    it.key.isNullable,
                    it.key.acceptedValues
                )
            }
            .toList()
    }

    override fun run(bot: Bot, paramsPropertyStore: BotPropertyStore): OperationResult {
        for (e in bot.botPropertyStore.entries) {
            if (paramsPropertyStore.containsByKey(e.key.key)) {
                e.setValue(paramsPropertyStore.getByKey(e.key.key))
            }
        }
        return OperationResult.ofObject(true, "Bot properties updated", bot.botPropertyStore
            .map { it.key.key to it.value }
            .toMap())
    }
}
